libecl/src/ecl_grid.c

```diff
@@ -1230,58 +1231,65 @@ static inline double tetrahedron_volume( tetrahedron_type tet ) {
  * inlining several operations, e.g. vector operations, and other tricks.
  */
 static double ecl_cell_get_signed_volume( ecl_cell_type * cell) {
-  ecl_cell_assert_center( cell );
+  if (GET_CELL_FLAG(cell , CELL_FLAG_VOLUME))
+    return cell->volume;
 
-  /*
-   * We make an activation record local copy of the cell's corners for less
-   * jumping in memory and better cache performance.
-   */
-  point_type center = cell->center;
-  point_type corners[ 8 ];
-  memcpy( corners, cell->corner_list, sizeof( point_type ) * 8 );
-
-  tetrahedron_type tet = { .p0 = center };
-  double           volume = 0;
-  /*
+  ecl_cell_assert_center( cell );
+  {
+    /*
+     * We make an activation record local copy of the cell's corners for less
+     * jumping in memory and better cache performance.
+     */
+    point_type center = cell->center;
+    point_type corners[ 8 ];
+    memcpy( corners, cell->corner_list, sizeof( point_type ) * 8 );
+    
+    tetrahedron_type tet = { .p0 = center };
+    double           volume = 0;
+    /*
       using both tetrahedron decompositions - gives good agreement
       with porv from eclipse init files.
-  */
-  /*
-   * The order of these loops is intentional and guided by profiling. It's much
-   * faster to access method, then the number, rather than the other way
-   * around. If you are to change this, please measure performance impact.
-   */
-  for( int method = 0; method < 2; ++method ) {
+    */
+
+    /*
+     * The order of these loops is intentional and guided by profiling. It's much
+     * faster to access method, then the number, rather than the other way
+     * around. If you are to change this, please measure performance impact.
+     */
+    for( int method = 0; method < 2; ++method ) {
       for( int itet = 0; itet < 12; ++itet  ) {
-          const int point0 = tetrahedron_permutations[ method ][ itet ][ 0 ];
-          const int point1 = tetrahedron_permutations[ method ][ itet ][ 1 ];
-          const int point2 = tetrahedron_permutations[ method ][ itet ][ 2 ];
-
-          tet.p1 = corners[ point0 ];
-          tet.p2 = corners[ point1 ];
-          tet.p3 = corners[ point2 ];
-          volume += tetrahedron_volume( tet ) / 6;
+	const int point0 = tetrahedron_permutations[ method ][ itet ][ 0 ];
+	const int point1 = tetrahedron_permutations[ method ][ itet ][ 1 ];
+	const int point2 = tetrahedron_permutations[ method ][ itet ][ 2 ];
+	
+	tet.p1 = corners[ point0 ];
+	tet.p2 = corners[ point1 ];
+	tet.p3 = corners[ point2 ];
+	volume += tetrahedron_volume( tet ) / 6;
```

Comment:
Perhaps we could change the name of `tetrahedron_volume` to something else if it doesn't actually compute the volume.